node/network/bitfield-distribution/src/lib.rs

```diff
@@ -443,6 +447,9 @@ where
 		}
 		one_per_validator.insert(validator.clone(), message.clone());
 
+		// If the peer has sent us a message, we don't need to send him the same.
+		job_data.message_sent_to_peer.entry(origin.clone()).or_default().insert(validator.clone());
```

Comment:
Isn't this already done above here? https://github.com/paritytech/polkadot/pull/2005/files#diff-6e99314d667011a73a36844509b05e6bfbf87e4cdcfa04d43d758d9ecf65fefbR424

Reply:
No, this is the set of received messages.

Reply:
Oh, good point. But why should we update the `message_sent_to_peer` if we've only received the message? It seems like we should only update the `message_received_from_peer`

Reply:
We should also check `message_received_from_peer` in the `interested_peers` computation in `relay_peers`, I guess.

Reply:
> Oh, good point. But why should we update the message_sent_to_peer if we've only received the message? It seems like we should only update the message_received_from_peer

If I have received the message from *you*, what is the benefit of sending you the same message again? This would happen below in `relay_message'. While you will not decrease my reputation for this, as I did not yet send you this message, I just waste my bandwidth for a message that is already known by you.

Reply:
> We should also check message_received_from_peer in the interested_peers computation in relay_peers, I guess.

We could also do it that way. I can change it that way if you like.

Reply:
> While you will not decrease my reputation for this, as I did not yet send you this message, I just waste my bandwidth for a message that is already known by you.

Who says they're the same message? We only know that they're from the same validator. The bitfield distribution section of the guide goes deeper into the rationale.

So I prefer the other change, if you don't mind doing it

Reply:
> Who says they're the same message? We only know that they're from the same validator. The bitfield distribution section of the guide goes deeper into the rationale.

That is a good point, but we actually don't account for this in the code? As far as I have seen, we only check if we have send/received a message based on the validator is message is coming from.

You are saying that we actually should look into the message and check if they are different, is that correct?

Reply:
> You are saying that we actually should look into the message and check if they are different, is that correct?

No, the change I suggest is just to check if we've received a message from the peer before sending a message. I think the thing this is trying to work around is where both peers send each other the message at the same time and then report each other.

Reply:
No, I don't try to work around this and this did not happen. We also just report a peer if it sends us the same message twice. We don't check if we send the message to this peer already for reporting the peer. Because as you said, this is very likely to end in a race condition.

However, I will do your requested changes.

Reply:
The other thing that I spoke about is that the code currently assumes that each validator only sends one bitfield per relay chain. Will this always be true or will we may change this in the future?

Reply:
I think this will be true for the foreseeable future